source/common/thread/thread_factory_singleton.cc

```diff
@@ -0,0 +1,18 @@
+#include "envoy/thread/thread.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Thread {
+
+ThreadFactory* ThreadFactorySingleton::thread_factory_{nullptr};
+
+// This function can not be inlined in the thread.h header due to the use of ASSERT() creating a
+// circular dependency with assert.h.
+void ThreadFactorySingleton::set(ThreadFactory* thread_factory) {
+  ASSERT((thread_factory == nullptr) != (thread_factory_ == nullptr));
```

Comment:
Sorry, this assert is an eyeful for me. Can you add a short comment on what you are testing?

Reply:
Done.